tests/websocket.c

```diff
@@ -8,319 +8,537 @@
 #include <pthread.h>
 #include <arpa/inet.h>
 #include <errno.h>
+#include <ctype.h>
+#include <getopt.h>
 
 #include <sys/types.h>
 #include <sys/socket.h>
 
 #include <event.h>
+#include <http_parser.h>
+
+#define DEBUG_LOGS 0
 
 struct host_info {
-	char *host;
-	short port;
+    char *host;
+    short port;
+};
+
+enum worker_state {
+    WS_INITIAL,
+    WS_SENT_HANDSHAKE,
+    WS_RECEIVED_HANDSHAKE,
+    WS_SENT_FRAME,
+    WS_COMPLETE
 };
 
 /* worker_thread, with counter of remaining messages */
 struct worker_thread {
-	struct host_info *hi;
-	struct event_base *base;
+    struct host_info *hi;
+    struct event_base *base;
+
+    int msg_target;
+    int msg_received;
+    int msg_sent;
+    int byte_count;
+    int id;
+    pthread_t thread;
+    enum worker_state state;
+    int timeout_seconds;
+
+    struct evbuffer *rbuffer;
+    int got_header;
+
+    struct evbuffer *wbuffer;
 
-	int msg_target;
-	int msg_received;
-	int msg_sent;
-	int byte_count;
-	pthread_t thread;
+    int verbose;
+    int fd;
+    struct event ev_r;
+    struct event ev_w;
 
-	struct evbuffer *buffer;
-	int got_header;
+    http_parser parser;
+    http_parser_settings settings;
 
-	int verbose;
-	struct event ev_w;
+    int (*debug)(const char *fmt, ...);
 };
 
+int debug_noop(const char *fmt, ...) {
+    (void)fmt;
+    return 0;
+}
+
+int debug_verbose(const char *fmt, ...) {
+    int ret;
+    va_list vargs;
+    va_start(vargs, fmt);
+    ret = vfprintf(stderr, fmt, vargs);
+    va_end(vargs);
+    return ret;
+}
+
+void
+hex_dump(struct worker_thread *wt, char *p, size_t sz) {
+    wt->debug("hex dump of %p (%ld bytes)\n", p, sz);
+    for (char *cur = p; cur < p + sz; cur += 16) {
+        char letters[16] = {0};
+        int limit = (cur + 16) > p + sz ? (sz % 16) : 16;
+        wt->debug("%08lx ", cur - p); /* address */
+        for (int i = 0; i < limit; i++) {
+            wt->debug("%02x ", (unsigned int)(cur[i] & 0xff));
+            letters[i] = isprint(cur[i]) ? cur[i] : '.';
+        }
+        for (int i = limit; i < 16; i++) { /* pad on last line */
+            wt->debug("   ");
```

Comment:
3 spaces for `%02x `? this could use a comment to explain the unusual string length.